src/package-linker.js

```diff
@@ -436,17 +436,34 @@ export default class PackageLinker {
     }
   }
 
-  determineTopLevelBinLinks(flatTree: HoistManifestTuples): Array<[string, Manifest]> {
+  determineTopLevelBinLinkOrder(flatTree: HoistManifestTuples): HoistManifestTuples {
     const linksToCreate = new Map();
-    for (const [dest, {pkg, isDirectRequire}] of flatTree) {
+    for (const [dest, hoistManifest] of flatTree) {
+      const {pkg, isDirectRequire} = hoistManifest;
       const {name} = pkg;
 
       if (isDirectRequire || (this.topLevelBinLinking && !linksToCreate.has(name))) {
-        linksToCreate.set(name, [dest, pkg]);
+        linksToCreate.set(name, [dest, hoistManifest]);
       }
     }
 
-    return Array.from(linksToCreate.values());
+    // Sort the array so that direct dependencies will be linked last.
+    // Bin links are overwritten if they already exist, so this will cause direct deps to take precedence.
+    // If someone finds this to be incorrect later, you could also consider sorting descending by
+    //   `a[1].level` which is the dependency tree depth. Direct deps will have level 0 and transitive
+    //   deps will have level > 0.
+    return Array.from(linksToCreate.values()).sort(([aDest, aHoistManifest], [bDest, bHoistManifest]): number => {
+      const aIsDirect = aHoistManifest.isDirectRequire;
+      const bIsDirect = bHoistManifest.isDirectRequire;
+
+      if (aIsDirect && !bIsDirect) {
+        return 1;
+      }
+      if (!aIsDirect && bIsDirect) {
+        return -1;
+      }
+      return 0;
+    });
```

Comment:
What do you think about splitting the array rather than doing a full sort? The algorithmic complexity would be better:

```
let transientBins = [], topLevelBins = [];
for (let ... of Array.from(linksToCreate.values())) {
    if (isDirectRequire) {
        topLevelBins.push();
    } else {
        transientBins.push();
    }
}

return [... transientBins, ... topLevelBins];
```

Reply:
I like that. The code is easier to read.